### ISSUE TITLE
Negative value for regex repeat produces unexpected error

### ISSUE THREAD

**Issue by markov2** [Perl/perl5#21555]
**Description**
An accidental negative repeat (in generated regexes) produces a cryptic error.   The regexp parser only understands positive integers, so produces the confusing report.

**Steps to Reproduce**
```perl
perl -we '"ab" =~ /.{-1,4}/;'
```
produces
```
Unescaped left brace in regex is passed through in regex; marked by <-- HERE in m/.{ <-- HERE -1,4}/ at -e line 1.
```

**Expected behavior**
It would be nice to have a better warning.  Have the parser accept negatives, but then produce an error like "do not use negative values in range".

**Perl configuration**
In 5.16, there is no error reported at all.  In 5.26, 5.36.1, and 5.38.0, I get this same error.

**Comment by jkeenan** [comment]
First trying the one-liner in various versions of `perlbrew`, then bisecting with this invocation ...
```
$ export ERR=/tmp/err/errors
$ perl Porting/bisect.pl \
--target=perl \
--start=v5.24.0 \
--end=v5.26.0 \
-e 'system(qq|./perl -we "q{ab} =~ /.{-1,4}/" 2>$ENV{ERR}|); die "See $ENV{ERR} for warnings thrown" if -s $ENV{ERR};'
```
... pointed to commit 8e84dec289 from 2016:
```
commit 8e84dec289732f088323c1bdc1e82f10841b362a
Author:     Karl Williamson <khw@cpan.org>
AuthorDate: Thu Jun 9 21:25:46 2016 -0600
Commit:     Karl Williamson <khw@cpan.org>
CommitDate: Fri Jun 17 12:45:20 2016 -0600

    Add missing deprecation message for unescaped '{' in regexes
    
    The use of literal '{' without being escaped has been deprecated since
    5.16, and warned on since 5.20.  In 5.24, this has been made illegal,
    with a bunch of CPAN modules broken by it, in spite of the long
    deprecation period.  See
    https://rt.perl.org/Ticket/Display.html?id=128139
    
    Unfortunately, I overlooked a code path, and not all instances that
    should have warned did so in fact.  This was spotted by Tom Wyant in
    https://rt.perl.org/Ticket/Display.html?id=128213
    
    This commit adds that warning, and rewords the fatal one slightly, and
    clarifies the whole thing in perldiag.
```
@khwilliamson, can you take a look?

**Comment by khwilliamson** [comment]
This is a legal pattern.  You may think of it as a failed quantifier, but someone not coming with that intention could see it as something completely different.  It is unfortunate that such a small detail could affect how the pattern is compiled.  The pattern syntax was designed for concision rather than clarity or error checking.  It's impossible in perl to change it much.  What that pattern compiles to hasn't changed, nor can it.  It matches any single character, and then the literal sequence `{-1,4}`. Until that warning was added, the pattern would compile silently with the result that if you were intending this to be a quantifier, it wasn't, and you might find out immediately, or not find out until much later, with years of subtle invalid results in between.  The warning now tells you how it is being treated, and how to fix it to avoid that warning if that is what you intended.  It was very painful getting that warning in; it had to be backed out, maybe even twice, because of the cpan breakage that first had to be fixed.  As a result, I would oppose any rewording.  But looking at perldiag, the text there could profitably be changed to stress that this can come about as a result of a quantifier gone bad.

**Comment by jkeenan** [comment]
> This is a legal pattern. You may think of it as a failed quantifier, but someone not coming with that intention could see it as something completely different. It is unfortunate that such a small detail could affect how the pattern is compiled. The pattern syntax was designed for concision rather than clarity or error checking. It's impossible in perl to change it much. What that pattern compiles to hasn't changed, nor can it. It matches any single character, and then the literal sequence `{-1,4}`. Until that warning was added, the pattern would compile silently with the result that if you were intending this to be a quantifier, it wasn't, and you might find out immediately, or not find out until much later, with years of subtle invalid results in between. The warning now tells you how it is being treated, and how to fix it to avoid that warning if that is what you intended. It was very painful getting that warning in; it had to be backed out, maybe even twice, because of the cpan breakage that first had to be fixed. As a result, I would oppose any rewording. But looking at perldiag, the text there could profitably be changed to stress that this can come about as a result of a quantifier gone bad.

@markov2, would you like to respond to @khwilliamson's analysis? Thanks.

**Comment by markov2** [comment]
I agree that it is a very good check: a useful restriction.  But harder than parsing the correct syntax, is producing optimal error messages for the many mistakes people can make.  In this case, it's even worse: you want people to write ```\{```, help them when they forget the ```{``` (in most cases correctly), but then lose some ways to detect other kinds of mistakes.  My error will especially appear primarily when regexes are (party) generated.

The parser could be changed to accept also negative integers, and then in a second step check that start and end
are positive.  In that case, a clear message can be produced.  Worth the performance loss at compilation?

Probably my main complaint, is that the error message "Unescaped left brace in regex is passed through in regex;" is incomprehensible for mortal souls like me.  "Is passed through" sounds like internal programmers' lingo.  It took me too long before the mistake became clear to me.  Maybe better something like "Unescaped left brace not interpreted as repeat".
